### DISCUSSION TITLE
Scoop search engine on the official website (https://scoop.sh/) is showing old and therefore incorrect software versions from my bucket!

### DISCUSSION THREAD

**Discussion by anderlli0053** [ScoopInstaller/Scoop#5440] [Help and Questions] [answered]
Scoop search engine on the official website (`https://scoop.sh/`) is showing old and therefore incorrect software versions from my bucket!

What is going on?? 

This keeps happening, but i cannot figure out why, is it the caching problem ??  But then it would be resolved in a matter of hours or maximum days... this now goes on for months...

**Comment by gpailler** [comment]
Hey @anderlli0053. The bucket you are talking about is https://github.com/anderlli0053/DEV-tools/tree/master/bucket, is that correct?

**Reply by anderlli0053** [reply]
Yes, that is correct

**Reply by anderlli0053** [reply]
@gpailler  Also the number of manifests in the the said bucket does not decrease nor increase, when i add custom manifests or remove them..

**Reply by gpailler** [reply]
There is a problem with the indexing of this repo indeed. Currently, the indexer has to clone the git repository and to iterate on all the commits to retrieve all the needed information for the search frontend. 
Your bucket as 10000+ manifests and 50000+ commits and it takes more than the maximum allowed duration (10 minutes) to process the whole repository.

I'm working on a refactoring of the indexer to index the repositories much faster. I hope it will be ready in the coming days. I'll keep you posted

**Reply by anderlli0053** [reply]
Thanks for taking a look into it!

**Reply by gpailler** [reply]
I applied some changes to the indexer and it processes buckets much faster now.

Unfortunately, the bucket https://github.com/anderlli0053/DEV-tools was added by @rasa in one of the exclusion list used by the indexer https://github.com/rasa/scoop-directory/blob/master/exclude.txt .

I can override this behavior and reindex https://github.com/anderlli0053/DEV-tools but I have 2 concerns with your repo:
- It contains more than 15,000 manifests and more than 5,000 commits today and it's growing very fast (+8300 the last month). Even with the optimizations I implemented, the indexer will fail to index your bucket again in the coming months because of the number of manifests/commits.
- More generally, we try to keep a clean index by removing buckets duplicating existing ones and I'm wondering if we should index the meta-buckets. The goal of the search engine is to search for applications in several buckets. If all the applications are in a single bucket, we don't need a search engine anymore 😄 . @rashil2000 @rasa do you remember if it was discussed already by the community?

**Reply by rasa** [reply]
I excluded https://github.com/anderlli0053/DEV-tools per https://github.com/rasa/scoop-directory/issues/95 as it has 16,151 manifests, and must be an "aggregate" bucket. But if @gpailler wants, I can un-exclude it. I would prefer not to, though. See rashil2000's [post below](https://github.com/ScoopInstaller/Scoop/discussions/5440#discussioncomment-5433538) re discussions regarding meta-buckets.

**Reply by rashil2000** [reply]
I too think we should exclude the meta-buckets. 

We discussed it previously in https://github.com/ScoopInstaller/ScoopInstaller.AzureFunctions/issues/7

**Reply by ZzZombo** [reply]
> There is a problem with the indexing of this repo indeed. Currently, the indexer has to clone the git repository and to iterate on all the commits to retrieve all the needed information for the search frontend. Your bucket as 10000+ manifests and 50000+ commits and it takes more than the maximum allowed duration (10 minutes) to process the whole repository.
> 
> I'm working on a refactoring of the indexer to index the repositories much faster. I hope it will be ready in the coming days. I'll keep you posted

Won't shallow clone work for the indexer to speed it up? What do you need the complete history for anyway?

**Reply by gpailler** [reply]
The indexer needs to fetch the complete history to index the commits associated with the manifests.

![image](https://user-images.githubusercontent.com/3621529/228718935-47b86700-4eef-48ce-8717-97a573338fed.png)

**Comment by anderlli0053** [comment]
@rasa @rashil2000  So OK, .. what's the manifest maximum limit for the bucket to not be excluded from various indexers??  

Talking about the regular indexer on the [official webpage](https://scoop.sh/#/apps?s=0&d=1&o=false) and the one [here](https://rasa.github.io/scoop-directory/)

**Reply by rashil2000** [reply]
There's no maximum limit. We just don't want to index aggregates/clones of other buckets - because of duplications in search results.

**Reply by anderlli0053** [reply]
But, is there no possibility to make the indexer filter the manifests that already exist in other buckets and just show the ones that are not in other buckets (especially official ones) ??

**Reply by gpailler** [reply]
Yes, it's an option I'm thinking of. We already have a filter for the official buckets, we can maybe add another filter for the duplicated manifests. But I need to check in more details how to properly detect duplicates and original manifests (manifest hash, commit date...)

**Comment by gpailler** [comment, marked as answer]
@anderlli0053 @rasa @rashil2000 
I finally managed to provide a solution for meta buckets indexing:
- New filter ["Distinct manifests only"](https://github.com/ScoopInstaller/scoopinstaller.github.io/issues/58) was added to https://scoop.sh
- The [indexer](https://github.com/ScoopInstaller/scoopinstaller.github.io-indexer) was massively refactored:
  - Performance improvements
  - Conversion from Azure functions to a Console application
  - Execution every 2 hours using GitHub actions to overcome the 10 minutes timeout and the costs on Azure
  - Detection of duplicated manifests based on name/version
  - Removal of the [exclusion list](https://github.com/ScoopInstaller/scoopinstaller.github.io-indexer/pull/18) used by the indexer
  - (+ kudos to Azure that was kind enough to increase the storage quota for the search service)

As a result:
- We have 1239 buckets in the index (1225 previously)
- We still have 5152 manifests from official buckets
- We have 27886 distinct manifests in the index
- We have a total 104619 manifests in the index (25527 previously)

**Reply by anderlli0053** [reply]
@gpailler 

Thank you very much for the update.
